server/repository/src/db_diesel/item_ledger.rs

```diff
@@ -0,0 +1,149 @@
+        invoice_number -> BigInt,
+        invoice_id -> Text,
+        reason -> Nullable<Text>,
+        invoice_status -> crate::db_diesel::invoice_row::InvoiceStatusMapping,
+        pack_size -> Double,
+        expiry_date -> Nullable<Date>,
+        batch -> Nullable<Text>,
+        cost_price_per_pack -> Double,
+        sell_price_per_pack -> Double,
+        total_before_tax -> Nullable<Double>,
+        number_of_packs -> Double,
+        type_precedence -> BigInt,
+        running_balance -> Double,
+    }
+}
+
+#[derive(Clone, Queryable, Debug, PartialEq, Default)]
+pub struct ItemLedgerRow {
+    pub id: String,
+    pub name: String,
+    pub item_id: String,
+    pub store_id: String,
+    pub quantity: f64,
+    pub datetime: NaiveDateTime,
+    pub invoice_type: InvoiceType,
+    pub invoice_number: i64,
+    pub invoice_id: String,
+    pub reason: Option<String>,
+    pub invoice_status: InvoiceStatus,
+    pub pack_size: f64,
+    pub expiry_date: Option<NaiveDate>,
+    pub batch: Option<String>,
+    pub cost_price_per_pack: f64,
+    pub sell_price_per_pack: f64,
+    pub total_before_tax: Option<f64>,
+    pub number_of_packs: f64,
+    pub type_precedence: i64,
+    pub running_balance: f64,
+}
+
+#[derive(Clone, Default)]
+pub struct ItemLedgerFilter {
+    pub store_id: Option<EqualFilter<String>>,
+    pub item_id: Option<EqualFilter<String>>,
+    pub datetime: Option<DatetimeFilter>,
+    pub invoice_type: Option<EqualFilter<InvoiceType>>,
+    pub invoice_status: Option<EqualFilter<InvoiceStatus>>,
+}
+
+impl ItemLedgerFilter {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn item_id(mut self, filter: EqualFilter<String>) -> Self {
+        self.item_id = Some(filter);
+        self
+    }
+
+    pub fn store_id(mut self, filter: EqualFilter<String>) -> Self {
+        self.store_id = Some(filter);
+        self
+    }
+
+    pub fn datetime(mut self, filter: DatetimeFilter) -> Self {
+        self.datetime = Some(filter);
+        self
+    }
+}
+
+pub struct ItemLedgerRepository<'a> {
+    connection: &'a StorageConnection,
+}
+
+impl<'a> ItemLedgerRepository<'a> {
+    pub fn new(connection: &'a StorageConnection) -> Self {
+        ItemLedgerRepository { connection }
+    }
+
+    pub fn count(&self, filter: Option<ItemLedgerFilter>) -> Result<i64, RepositoryError> {
+        let query = create_filtered_query(filter);
+        Ok(query
+            .count()
+            .get_result(self.connection.lock().connection())?)
+    }
+
+    pub fn query(
+        &self,
+        pagination: Pagination,
+        filter: Option<ItemLedgerFilter>,
+    ) -> Result<Vec<ItemLedgerRow>, RepositoryError> {
+        let mut query = create_filtered_query(filter);
+        query = query
+            .order(item_ledger::datetime.desc())
+            .then_order_by(item_ledger::id.desc())
+            .then_order_by(item_ledger::type_precedence.asc());
```

Comment:
Need sorting to match with view so that the balances to match with the correct entry 👀